### PR TITLE
Metrics interfaces

### PR DESCRIPTION
**Pull request by SergeyKanzhelev**
#24 

Interfaces only for metrics reporting. There is no way to instantiate it yet to start reporting those metrics.

### CODE REVIEW

File: src/OpenTelemetry.Abstractions/Metrics/Implementation/IMetricBuilder{T}.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        /// <param name="unit">Unit of the metric.</param>
+        /// <returns>This builder instance.</returns>
+        IMetricBuilder<T> SetUnit(string unit);
+
+        /// <summary>
+        /// Sets the description of the <see cref="IMetric{T}"/>.
+        /// </summary>
+        /// <param name="labelKeys">List of keys for dynamic labels.</param>
+        /// <returns>This builder instance.</returns>
+        IMetricBuilder<T> SetLabelKeys(List<LabelKey> labelKeys);
```
Comment (z1c0) [inline src/OpenTelemetry.Abstractions/Metrics/Implementation/IMetricBuilder{T}.cs]: Other methods so far use `IEnumerable` . Should we use that here as well?

File: src/OpenTelemetry.Abstractions/Metrics/Implementation/IMetricBuilder{T}.cs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+        /// <summary>
+        /// Sets the description of the <see cref="IMetric{T}"/>.
+        /// </summary>
+        /// <param name="constantLabels">Set of labels with values.</param>
+        /// <returns>This builder instance.</returns>
+        IMetricBuilder<T> SetConstantLabels(Dictionary<LabelKey, string> constantLabels);
```
Comment (z1c0) [inline src/OpenTelemetry.Abstractions/Metrics/Implementation/IMetricBuilder{T}.cs]: Maybe `IDictionary`?

File: src/OpenTelemetry.Abstractions/Metrics/IMeasurement.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+namespace OpenTelemetry.Metrics
+{
+    /// <summary>
+    /// Immutable representation of a measurement.
+    /// </summary>
+    public class IMeasurement
```
Comment (z1c0) [inline src/OpenTelemetry.Abstractions/Metrics/IMeasurement.cs]: This should be `interface`

Comment (SergeyKanzhelev) [reply src/OpenTelemetry.Abstractions/Metrics/IMeasurement.cs]: thank you!

File: src/OpenTelemetry.Abstractions/Metrics/IMetric{T}.cs [hunk 4]
```diff
@@ -0,0 +1,72 @@
+﻿// <copyright file="IMetric{T}.cs" company="OpenTelemetry Authors">
```
Comment (lmolkova) [inline src/OpenTelemetry.Abstractions/Metrics/IMetric{T}.cs]: is it the new way of naming files with generics?

File: src/OpenTelemetry.Abstractions/Metrics/Implementation/MeasureType.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+// </copyright>
+
+namespace OpenTelemetry.Metrics.Implementation
+{
+    public enum MeasureType
+    {
+        /// <summary>
+        /// Long type values.
+        /// </summary>
+        LONG,
```
Comment (lmolkova) [inline src/OpenTelemetry.Abstractions/Metrics/Implementation/MeasureType.cs]: Pascal case?

File: src/OpenTelemetry.Abstractions/Metrics/IGaugeLong.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+namespace OpenTelemetry.Metrics
+{
+    using OpenTelemetry.Metrics.Implementation;
+
+    /// <summary>
+    /// Gauge metric, to report instantaneous measurement of a double value. Cumulative values can go
+    /// up or stay the same, but can never go down.Cumulative values cannot be negative.
+    /// </summary>
+    public interface IGaugeLong : IMetric<IGaugeLongTimeSeries>
+    {
```
Comment (lmolkova) [inline src/OpenTelemetry.Abstractions/Metrics/IGaugeLong.cs]: Java implementation is different

https://github.com/open-telemetry/opentelemetry-java/blob/master/api/src/main/java/io/opentelemetry/metrics/GaugeLong.java

Is it intended?

File: src/OpenTelemetry.Abstractions/Metrics/IMetric{T}.cs (hunk 4 above)
Comment (SergeyKanzhelev) [reply src/OpenTelemetry.Abstractions/Metrics/IMetric{T}.cs]: yep. Stylecop asks for it

**Pull request merged**